### DISCUSSION TITLE
Recommended solution for multi-tenancy?

### DISCUSSION THREAD

**Discussion by efenderbosch** [vladimir-bukhtoyarov/bucket4j#15] [Q&A] [answered]
We'll be making various 3rd party API calls. Some will be with our API credentials and some will be on behalf of our customers using their API credentials. For calls using our client's API credentials, they should each have separate buckets.

For local buckets, should we just store the buckets in a ``ConcurrentHashMap`` keyed by the tenant's ID, or is there a more efficient solution? Generally the API limits will be the same for all tenants, so sharing a ``BucketConfiguration`` across all instances of ``LockFreeBucket`` would save memory.

For JCache, we'd simply want to use the tenant's ID as the key in JCache, correct?

**Comment by vladimir-bukhtoyarov** [comment]
I suppose ```ConcurrentHashMap``` would be good general solution to storing local bucket. To introduce solution with lesser contention it is need to know something special about your data, for example if count of "tenant" is finite and known then your can just use open addressing and store buckets in the array.


### ----------------------------------------------------------------------------
> Generally the API limits will be the same for all tenants, so sharing a BucketConfiguration across all instances of LockFreeBucket would save memory.

Reusage of configuration is not supported as first class citizen in the builder API, but your goal can be achieved in following way:
```java
BucketConfiguration sharedConfiguration = Bucket4j.builder()
           .addLimitedBandwidth(100, Duration.ofMinute(1))
           .createConfiguration();

Bucket bucket1 = new  LockFreeBucket(sharedConfiguration);
Bucket bucket2 = new  LockFreeBucket(sharedConfiguration);
Bucket bucket3 = new  LockFreeBucket(sharedConfiguration);
```
Or like this:
```java
Bucket bucket1 = Bucket4j.builder()
           .addLimitedBandwidth(100, Duration.ofMinute(1))
           .build();
BucketConfiguration sharedConfiguration = bucket1.getConfiguration();

Bucket bucket2 = new  LockFreeBucket(sharedConfiguration);
Bucket bucket3 = new  LockFreeBucket(sharedConfiguration);
```
### ----------------------------------------------------------------------------
> For JCache, we'd simply want to use the tenant's ID as the key in JCache, correct?

Exactly.

**Comment by vladimir-bukhtoyarov** [comment]
Looks like question fully answered

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
There is better way introduced in scope of ([JHipster spapport](https://github.com/vladimir-bukhtoyarov/bucket4j/issues/26)). [ProxyManager](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/doc-pages/jcache-usage.md#example-1---limiting-access-to-http-server-by-ip-address) abstraction provides ability to avoid handy management of bucket instances.
